src/pybind/mgr/mgr_util.py

```diff
@@ -23,6 +27,256 @@
+                        -errno.ENOENT, "Volume '{0}' not found".format(fs_name))
+                raise CephfsConnectionException(-e.args[0], e.args[1])
+            self.connections[fs_name] = conn
+            return conn.get_fs_handle()
+
+    def put_fs_handle(self, fs_name):
+        with self.lock:
+            conn = self.connections.get(fs_name, None)
+            if conn:
+                conn.put_fs_handle(notify=lambda: self.cond.notifyAll())
+
+    def _del_fs_handle(self, fs_name, wait=False):
+        conn = self.connections.pop(fs_name, None)
+        if conn:
+            conn.del_fs_handle(
+                waiter=None if not wait else lambda: self.cond.wait())
+
+    def del_fs_handle(self, fs_name, wait=False):
+        with self.lock:
+            self._del_fs_handle(fs_name, wait)
+
+    def del_all_handles(self):
+        with self.lock:
+            for fs_name in list(self.connections.keys()):
+                logger.info("waiting for pending ops for '{}'".format(fs_name))
+                self._del_fs_handle(fs_name, wait=True)
+                logger.info("pending ops completed for '{}'".format(fs_name))
+            # no new connections should have been initialized since its
+            # guarded on shutdown.
+            assert len(self.connections) == 0
+
+
+class CephfsClient(object):
+    def __init__(self, mgr):
+        self.mgr = mgr
+        self.stopping = Event()
+        self.connection_pool = ConnectionPool(self.mgr)
+
+    def shutdown(self):
+        logger.info("shutting down")
+        # first, note that we're shutting down
+        self.stopping.set()
+        # second, ask purge threads to quit
+        self.purge_queue.cancel_all_jobs()
+        # third, delete all libcephfs handles from connection pool
+        self.connection_pool.del_all_handles()
+
+    def get_fs(self, fs_name):
+        fs_map = self.mgr.get('fs_map')
+        for fs in fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs
+        return None
+
+    def get_mds_names(self, fs_name):
+        fs = self.get_fs(fs_name)
+        if fs is None:
+            return []
+        return [mds['name'] for mds in fs['mdsmap']['info'].values()]
+
+
+class FSHandle(object):
+    def __init__(self, name, handle):
+        self.name = name
+        self.handle = handle
+
+    def __repr__(self):
+        return self.name
+
+
+def connection_pool_wrap(func):
+    """
+    decorator that wraps CephfsClient calls by fetching filesystem handle
+    from the fs_name.
+    """
+    def conn_wrapper(self, fs_name, *args, **kwargs):
+        # note that force arg is available for remove type commands
+        force = kwargs.get('force', False)
+
+        if self.stopping.isSet():
+            return -errno.ESHUTDOWN, "", "shutdown in progress"
+
+        # fetch the connection from the pool
+        try:
+            fs_h = self.connection_pool.get_fs_handle(fs_name)
```

Comment:
At first look, this seemed incorrect since `connection_pool` (and `stopping` variable above) are a member of `CephfsClient` class. But I guess implementations would derive from `CephfsClient` class. Any thoughts about making this decorator a part of `CephfsClient` class?

Reply:
Right I think it makes sense as a static member.
Though I see that the volumes implementation just moved to a contextmanager approach. Does it make sense to pull this down to here?

Reply:
Yeh -- I found using context manager in mgr/volumes very convenient. We should include that in mgr utils.